### DISCUSSION TITLE
Install Mastodon docker-compose (help me)

### DISCUSSION THREAD

**Discussion by collex100** [mastodon/mastodon#31678] [Q&A]
**OS: Ubuntu 22.04**

The command to eliminate the drop-down message in Ubuntu 22.04
```
sudo apt-get purge needrestart
```
Installation Docker-compose (my installation script)
```
curl -s https://raw.githubusercontent.com/6Ministers/Mastodon-docker-compose/master/setup.sh | sudo bash -s
```
Go to the catalog opt
```
cd /opt
```
Cloning the repository
```
git clone https://github.com/mastodon/mastodon.git
```
Go to the catalog mastodon
```
cd mastodon
```
Specify the active branch of the repository
```
git checkout $(git tag -l | grep -v 'rc[0-9]*$' | sort -V | tail -n 1)
```
Creating a working variable file
```
cp .env.production.sample .env.production
```
Creating an assembly
```
docker-compose build
```
Creating a key SECRET_KEY_BASE
```
SECRET_KEY_BASE=$(docker-compose run --rm web bundle exec rails secret)
```
Writing the key to .env.production
```
sed -i -e "s/SECRET_KEY_BASE=/&$SECRET_KEY_BASE/" .env.production
```
Creating a key OTP_SECRET
```
OTP_SECRET=$(docker-compose run --rm web bundle exec rails secret)
```
Writing the key to .env.production
```
sed -i -e "s/OTP_SECRET=/&$OTP_SECRET/" .env.production
```
Creating keys vapid_key
```
docker-compose run --rm web bundle exec rails mastodon:webpush:generate_vapid_key
```
Writing keys to .env.production
Copying manually from the console

Domain entry in .env.production
LOCAL_DOMAIN=example.com (I specify my domain)

Re-creating the assembly
```
docker-compose build
```
It is not clear which way to go next, there are two of them and each gives out its own mistake.

**Path 1**

Set up the database
```
docker-compose run --rm web rails db:migrate
```
A recommendation appears in the console to create the necessary keys
```
docker-compose run --rm web bin/rails db:encryption:init
```
I manually copy the keys from the console to the variables file '.env.production'
Pre-compile Mastodon’s assets.
This command below causes an error
```
docker-compose run --rm web rails assets:precompile
```
**ERROR**
```
[+] Creating 2/0
 ✔ Container mastodon-redis-1  Running                                     0.0s
 ✔ Container mastodon-db-1     Running                                     0.0s
bin/rails aborted!
Errno::EACCES: Permission denied @ rb_sysopen - /opt/mastodon/public/assets/.manifest.json (Errno::EACCES)

Tasks: TOP => assets:precompile
(See full trace by running task with --trace)
```

This command below does not help solve the access issue.
```
docker-compose run --rm -u root web chown -R mastodon:mastodon public/assets public/packs public/system
```

Next, you need to run docker-compose
```
docker-compose up -d
```
According to the instructions, you need to run docker-compose, but the containers do not start, but get into the status of eternal restart.

**Path 2**
User Settings
Causes an error with database access
```
docker compose run --rm web bundle exec rails mastodon:setup
```

**ERROR**
```
PostgreSQL host: /var/run/postgresql
PostgreSQL port: 5432
Name of PostgreSQL database: mastodon_production
Name of PostgreSQL user: mastodon
Password of PostgreSQL user:
Database connection could not be established with this configuration, try again.
connection to server on socket "/var/run/postgresql/.s.PGSQL.5432" failed: No such file or directory
        Is the server running locally and accepting connections on that socket?
```

Next, you need to run docker-compose
```
docker-compose up -d
```

Taking into account that path 1 and path 2 do not work correctly, then launching docker-compose brings the result that containers cannot start, they are constantly being restarted.

![Снимок экрана 2024-09-01 120820](https://github.com/user-attachments/assets/eddeb336-8710-4163-a5f3-51b2d4120e73)

**Path 3 - launch [from source](https://docs.joinmastodon.org/admin/install/),It didn't give any results either.**
I get a white screen in the browser and the Mastodon logo in the upper left corner and a lot of errors in the browser console that I can't access other content. I did not succeed in launching more than one way. It's very strange. ((

Please help me.

**Comment by hidetoshi-tamaki** [comment]
Hello,

I encountered the same issue as you in path 3. However, the documentation (Installing from source) was updated in October 2024, after the release of v4.3.0, and the following paragraph was added:

_Allow other users to traverse the mastodon user’s home directory so that nginx’s www-data user can access asset files:_
`chmod o+x /home/mastodon`

In my case, the issue you mentioned (Mastodon logo in the upper left corner and many errors in the browser console) was resolved by this command. However, I’m still experiencing the same issue in Path 1 after running the following command:

```
$ sudo docker-compose run --rm web rails assets:precompile
[+] Creating 2/0
 ✔ Container mastodon-redis-1  Running                                                                                                                                                                         0.0s 
 ✔ Container mastodon-db-1     Running                                                                                                                                                                         0.0s 
bin/rails aborted!
Errno::EACCES: Permission denied @ rb_sysopen - /opt/mastodon/public/assets/.manifest.json (Errno::EACCES)

Tasks: TOP => assets:precompile
(See full trace by running task with --trace)
```

Thanks!

**Reply by rriemann** [reply]
> Errno::EACCES: Permission denied @ rb_sysopen - /opt/mastodon/public/assets/.manifest.json (Errno::EACCES)
Could you solve it since then?

**Reply by hidetoshi-tamaki** [reply]
Hi @rriemann ! Unfortunately, I couldn't solve it. So, I decided to continue using it with a non-Docker configuration.

**Comment by AdrianPaulCarrieres** [comment]
Same issue here since upgrading to 4.3...

**Reply by rriemann** [reply]
Do you have selinux? For me, it was this and ordinary Unix permissions. I got it working with rootless podman eventually.

**Reply by AdrianPaulCarrieres** [reply]
Nope, classic ubuntu and docker...

**Comment by mastoca** [comment]
This is still broken and unanswered. 

Is this option being removed from official support? If so, it should be clearly stated so others don't waste their time with this. 

Or should be an issue instead of a discussion.

**Reply by AdrianPaulCarrieres** [reply]
You apparently and actually don’t need it in docker. Not sure why it was more or less working before the 4.3 versions, but I just ignored that step and let the container handle it without any issue

**Reply by mastoca** [reply]
Hi @AdrianPaulCarrieres, I sorry I don't understand what you mean by 'let the container handle it'
the public/system directory is empty and my assumption is the precompile command is supposed to copy the packs etc?

**Reply by rriemann** [reply]
The docker image has it pre-compiled. It could be that a volume is overwriting it. So let's doublecheck:

~~~
$ podman run --user 0 --rm -it ghcr.io/mastodon/mastodon:v4.3.2 bash
root@19fdea890939:/opt/mastodon# ls -lh /opt/mastodon/public
total 152K
lrwxrwxrwx. 1 root     root       15 Dec  3 15:27 500.html -> assets/500.html
drwxr-xr-x. 1 root     root     1.3K Dec  3 15:34 assets
drwxr-xr-x. 1 root     root       16 Dec  3 15:27 avatars
-rw-r--r--. 1 root     root     3.0K Dec  3 15:27 badge.png
-rw-r--r--. 1 root     root     3.5K Dec  3 15:27 embed.js
drwxr-xr-x. 1 root     root     143K Dec  3 15:27 emoji
drwxr-xr-x. 1 root     root       16 Dec  3 15:27 headers
drwxr-xr-x. 1 root     root       18 Dec  3 15:27 ocr
-rw-r--r--. 1 root     root      94K Dec  3 15:27 oops.gif
-rw-r--r--. 1 root     root      17K Dec  3 15:27 oops.png
drwxr-xr-x. 1 root     root      246 Dec  3 15:34 packs
-rw-r--r--. 1 root     root      190 Dec  3 15:27 robots.txt
drwxr-xr-x. 1 root     root       32 Dec  3 15:27 sounds
lrwxrwxrwx. 1 root     root       11 Dec  3 15:27 sw.js -> packs/sw.js
lrwxrwxrwx. 1 root     root       15 Dec  3 15:27 sw.js.map -> packs/sw.js.map
drwxr-xr-x. 1 mastodon mastodon    0 Dec  3 15:36 system
-rw-r--r--. 1 root     root     1.1K Dec  3 15:27 web-push-icon_expand.png
-rw-r--r--. 1 root     root      721 Dec  3 15:27 web-push-icon_favourite.png
-rw-r--r--. 1 root     root      590 Dec  3 15:27 web-push-icon_reblog.png
~~~

See also: https://github.com/mastodon/mastodon/issues/32331

**Reply by mastoca** [reply]
My directory looks identical to that. However, none of the files for example in /packs load in the browser.

**Reply by rriemann** [reply]
Can you describe the flow that you envision when people call an URL to download the asset files? You have a reverse proxy? In our outside of the container? Which one?

**Reply by mastoca** [reply]
I am using nginx with proxypass to port 3000. Nginx is on host outside of all containers. Note that some assets load such as the PNGs and ico files. However anything that's compiled (css and js) are not loading.

**Reply by rriemann** [reply]
With the little information on your setup at hand, I would guess that with nginx outside of the container, you cannot easily point the file root of nginx for your domain to the files inside the container. The idea of the docker-compose setup promoted by Mastodon is to serve the assets files with rails (which is inefficient usually) to ngnix and then let nginx cache those files (which is then again efficient).

To permit rails serving files from /public, you need to tell rails explicitly. This is documented here:

https://docs.joinmastodon.org/admin/config/#rails_serve_static_files

You have two more options:

1. you copy the asset files outside of the container after each update into a directory of the host that nginx can reach. You have to copy the assets after each container image update.
2. you have nginx inside the docker network and configure a named volume to /opt/mastodon/public in mastodon that you then also mount in parallel to your nginx container so that nginx has access. The named volume must be explictly emptied when you update to a new container image version. This approach is demonstrated in this otherwise outdated blog post: https://blog.riemann.cc/digitalisation/2022/02/09/mastodon-setup-with-docker-and-caddy/

**Reply by mastoca** [reply]
@rriemann an excellent response. Thank you. This now clears up the rather large gap in my understanding/expectation of a docker compose setup. I guess there's a desire to give options on how the static assets should be served to avoid rails doing it. Now that I understand I can solve this and finish my setup. But speaking for others, I'm not sure this is a welcoming design. IMO, the ball is definitely getting dropped here.

We should explicitly say that the docker-compose file is a base/incomplete solution and provide, like some other projects do, a few slightly opinionated yaml files eg. `docker-compose-caddy.yaml` or `docker-compose-nginx.yaml` I will design something for me and share my final result and hope this will help the many that have struggled and gave up...to perhaps try again.

**Comment by loafykitten** [comment]
Popping in to say I've also had this problem. Trying to enter the docker container via root and run the precompile doesn't work either as it complains things like node aren't installed.

I've tried solving this issue for hours and sadly I just can't get it solved. Mastodon has really dropped the ball when it comes to docker documentation.

**Reply by rriemann** [reply]
It should have all assets. Check my reply at https://github.com/mastodon/mastodon/discussions/31678#discussioncomment-11845282

**Reply by mastoca** [reply]
The problem I encountered and couldn't fix is that the assets aren't being served at all. eg. all files in /packs fail. How did you solve that problem? I think that's where people are getting stuck and trying to recompile because in the system directory there is nothing.

**Reply by AdrianPaulCarrieres** [reply]
It's because the system directory is empty, you should check the public one directly

**Comment by wolfspyre** [comment]
WOW I feel less dumb now. 

I've spent... WEEKS trying to sort this out. (amongst other things)..  and reading thru this has me feeling.... well.... at least in good company ;) 

@rriemann  - is the `/opt/mastodon/public` directory the only content directory to extract from a build? 
(I assume web?) Should anything be extracted from the streaming server container? 

I'm looking into how to make a gitlab release artifact with the contents of the relevant folders,alongside my homebuilt containers, 
 such that I've got an asset bundle I can feed an adjacent webserver the bits to cache ... 


The shared folder mount pattern was **really misleading** to me... 
The docs need updating, IMO as there's not a good happy path rn from a blank slate. 


(...or ... alternatively.... perhaps I'm just horrible at following instructions ?? )

**Reply by mastoca** [reply]
@wolfspyre welcome to the other side. I too almost gave up. I wonder how many DNF'd...

I've been `docker cp <container id>:/opt/mastodon/public public/` to keep the folders 'populated' after updating.

**Reply by wolfspyre** [reply]
@mastoca  I feel seen ;) thanks ;) 

it's been a good exercise tho ... TBH this was about as hard + / - as getting a matrix / synapse server functional ... but I'm BARELY on the other side... I just finally got a page to render from the stack..... 

- still haven't bolted in ES.. 
- still haven't set up any accounts
- still haven't gotten PG multi-node
- still haven't set up redis multi-node
- still haven't formalized nginx config ... 
- still haven't decided on how I wanna reverse-reverse proxy things 
    - ( (opnsense fw pair running haproxy)  -> [[ qemu vm [ nginx -> [docker containers [ web || sidekiq || streaming ] ]  ]] )

fun!

**Reply by rriemann** [reply]
Hi all! So I only extract this folder currently. However, I noticed that this means that all assets that are not outside the container and cannot be found by caddy will still be served from within the container – thus have a strain on the ruby/rails app. This is suboptimal.

I then made an attempt to measure the difference between loading images just via caddy and loading them reverse-proxied through ruby/rails and found in my otherwise idle server no significant performance drawback. So I left it then this way (images are reverse proxied).

In the meanwhile, I have learnt a few more things:
- many containers (synapse, postgresql, redis) can be excuted with arbitary user id and create their volume (mount) files with that ID
- named volumes are just created somewhere on the host in a folder hard to find
- if you would use a named volume and use user id 'www-run' or alike to run mastodon, then you could point caddy to that folder hard to find that contains everything in the named volume.

**Reply by wolfspyre** [reply]
@rriemann 
thanks for throwing me a clue! it’s appreciated

yeah, 

your discovery is apt; but I do wonder as to the ultimate benefit of pulling some content out of the container here … (as in, is the value gained worth the  extra complexity )

I suspect that the project would benefit from a more complete diagram of how the pieces are intended to fit together … 

perhaps a github issue isn’t the best place for this conversation; but barring a more apt location being presented;  this is better than nothing…

there are dozens of totally legitimate ways to get rolling… 
that’s okay; and to be expected ….

What I have struggled with is that the canonical mastodon project, as is, doesn’t seem to have a complete working exemplar…  (as far as i can tell) 
and the documentation on the ‘setting up your server’ page sorta glosses over the initial setup steps…. (as is easy to do when you’ve done it dozens s of times)

how “the webserver doing reverse proxying” should get a copy of the content is unclear, and my confusion is exacerbated by a plethora of legitimate but orthogonal patterns of putting things together…. 

I wonder if anyone’s put together a diagram in ilograph, d2, mermaid, or python diagrams  (diagrams-as-code) of how the stack works…

and if there’s an almost-sane way to assess the state of a rig and help point out what ISNT right in a fashion that’s not either a yak-shaving exercise, or another chapter in adventures-in-ocean-boiling …


to your point about docker volumes; check out 
[local-persist](https://github.com/MatchbookLab/local-persist) 

i’ve been using it for years. works great.

**Comment by AdrianPaulCarrieres** [comment]
Okay so, I migrated twice and started from scratch both times as well. I never had anything to do about the public directory, only one thing on the elasticsearch directory for permissions. 

Basically, you take the docker-compose.yml file, the .env.production file, the nginx-config file, change the values, copy the nginx file and do the symbolic link and then `docker compose up -d`, lets docker create the different volumes and mounted folder and you should be done.

To be noted that I'm using a S3 to serve user uploaded files and stuff like that.

My nginx configuration looks like that:

```
map $http_upgrade $connection_upgrade {
  default upgrade;
  ''      close;
}

upstream backend {
    server 127.0.0.1:3000 fail_timeout=0;
}

upstream streaming {
    # Instruct nginx to send connections to the server with the least number of connections
    # to ensure load is distributed evenly.
    least_conn;

    server 127.0.0.1:4000 fail_timeout=0;
    # Uncomment these lines for load-balancing multiple instances of streaming for scaling,
    # this assumes your running the streaming server on ports 4000, 4001, and 4002:
    # server 127.0.0.1:4001 fail_timeout=0;
    # server 127.0.0.1:4002 fail_timeout=0;
}

proxy_cache_path /var/cache/nginx levels=1:2 keys_zone=CACHE:10m inactive=7d max_size=1g;

server {
  listen 80;
  listen [::]:80;
  server_name example.com;
  root /home/mastodon/mastodon/public;
  location /.well-known/acme-challenge/ { allow all; }
  location / { return 301 https://$host$request_uri; }
}

server {
  listen 443 ssl http2;
  listen [::]:443 ssl http2;
  server_name example.com;

  ssl_protocols TLSv1.2 TLSv1.3;

  # You can use https://ssl-config.mozilla.org/ to generate your cipher set.
  # We recommend their "Intermediate" level.
  ssl_ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:DHE-RSA-CHACHA20-POLY1305;

  ssl_prefer_server_ciphers on;
  ssl_session_cache shared:SSL:10m;
  ssl_session_tickets off;

  # Uncomment these lines once you acquire a certificate:
  ssl_certificate     /home/mastodon/.acme.sh/example.com_ecc/fullchain.cer;
  ssl_certificate_key /home/mastodon/.acme.sh/example.com_ecc/example.com.key;

  keepalive_timeout    70;
  sendfile             on;
  client_max_body_size 99m;

  root /home/mastodon/mastodon/public;

  gzip on;
  gzip_disable "msie6";
  gzip_vary on;
  gzip_proxied any;
  gzip_comp_level 6;
  gzip_buffers 16 8k;
  gzip_http_version 1.1;
  gzip_types text/plain text/css application/json application/javascript text/xml application/xml application/xml+rss text/javascript image/svg+xml image/x-icon;
  gzip_static on;

  location / {
    try_files $uri @proxy;
  }

  # If Docker is used for deployment and Rails serves static files,
  # then needed must replace line `try_files $uri @proxy;` with `try_files $uri @proxy;`.
  location = /sw.js {
    add_header Cache-Control "public, max-age=604800, must-revalidate";
    add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";
    try_files $uri @proxy;
  }

  location ~ ^/assets/ {
    add_header Cache-Control "public, max-age=2419200, must-revalidate";
    add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";
    try_files $uri @proxy;
  }

  location ~ ^/avatars/ {
    add_header Cache-Control "public, max-age=2419200, must-revalidate";
    add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";
    try_files $uri @proxy;
  }

  location ~ ^/emoji/ {
    add_header Cache-Control "public, max-age=2419200, must-revalidate";
    add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";
    try_files $uri @proxy;
  }

  location ~ ^/headers/ {
    add_header Cache-Control "public, max-age=2419200, must-revalidate";
    add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";
    try_files $uri @proxy;
  }

  location ~ ^/packs/ {
    add_header Cache-Control "public, max-age=2419200, must-revalidate";
    add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";
    try_files $uri @proxy;
  }

  location ~ ^/shortcuts/ {
    add_header Cache-Control "public, max-age=2419200, must-revalidate";
    add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";
    try_files $uri @proxy;
  }

  location ~ ^/sounds/ {
    add_header Cache-Control "public, max-age=2419200, must-revalidate";
    add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";
    try_files $uri @proxy;
  }

  location ~ ^/system/ {
    add_header Cache-Control "public, max-age=2419200, immutable";
    add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";
    add_header X-Content-Type-Options nosniff;
    add_header Content-Security-Policy "default-src 'none'; form-action 'none'";
    try_files $uri @proxy;
  }

  location ^~ /api/v1/streaming {
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header Proxy "";

    proxy_pass http://streaming;
    proxy_buffering off;
    proxy_redirect off;
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection $connection_upgrade;

    add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";

    tcp_nodelay on;
  }

  location @proxy {
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header Proxy "";
    proxy_pass_header Server;

    proxy_pass http://backend;
    proxy_buffering on;
    proxy_redirect off;
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection $connection_upgrade;

    proxy_cache CACHE;
    proxy_cache_valid 200 7d;
    proxy_cache_valid 410 24h;
    proxy_cache_use_stale error timeout updating http_500 http_502 http_503 http_504;
    add_header X-Cached $upstream_cache_status;

    tcp_nodelay on;
  }

  error_page 404 500 501 502 503 504 /500.html;
}
```

**Reply by wolfspyre** [reply]
Hai @AdrianPaulCarrieres !! 

:) Thanks for that! 

Are you running nginx on the host directly?
Or as another container in the mastodon stack?

the compose file in the repo doesn't have a frontend webserver component, and the canonical guidance on implementation is sorta ... handwavey...
 
(or, perhaps I'm just really bad at reading ... which is arguably equally likely... )

**Reply by AdrianPaulCarrieres** [reply]
I run nginx on the host directly!

**Reply by AdrianPaulCarrieres** [reply]
Also for anyone reading this, if you’re using a special theme (like tangerine ui) don’t forget to update it once you get to mastodon 4.4, that was an issue of mine (totally on my side, not a problem from tangerine :))

**Reply by wolfspyre** [reply]
> I run nginx on the host directly!

Yeah.... that's what I ended up doing after many dif contortions.... tried using nginx-proxy-manager, but REALLY DID NOT LIKE the opinionated tooling there for this use case.... 
  - no way to disable the 'helpful' addition of **EVERY POSSIBLE LB IPS FROM ALL CLOUD PROVIDERS** as trusted proxy addrs... 
  - no happypath way to add your **OWN** tls certs... it **ONLY** supports acme, (and near as I can tell only letsencrypt but that may not be accurate) 
  - no happypath way to have the nginxhost serve content itself... 
  - no proxyprotocol support
  
  then tried adding a generic nginx container but that was kinda messy and was when I realized that there was something drastically divergent with my understanding of how the sausage is made there. 
  
  it's a nice bit of kit... just tooled for a use case not mine.

**Reply by AdrianPaulCarrieres** [reply]
Yeah, at first I wanted nginx as a container as well, but I figured I would make this more quickly :)
